gcs_light_client.mediawiki

```diff
@@ -0,0 +1,992 @@
+    write_bits_big_endian(stream, n, k)
+</pre>
+
+<pre>
+golomb_decode(stream, k) -> int:
+    let c = stream.read_bit()
+
+    let n = 0
+    while c == 0:
+        n++
+        c = stream.read_bit()
+
+    let r = b.read_bits_big_endian(k)
+
+    where read_bits_big_endian(k) decodes a fixed-length big-endian integer of
+        k-bits 
+
+    c*m + r
+</pre>
+
+To aide in understanding we provide the following examples of using Golomb-Rice
+encoding to code integers given <code>m=5</code>
+<pre>
+n  = (q, r) = c
+0  = (0, 0) = 0 00
+1  = (0, 1) = 0 01
+2  = (0, 2) = 0 10
+3  = (0, 3) = 0 110
+4  = (0, 4) = 0 111
+5  = (1, 0) = 10 00
+6  = (1, 1) = 10 01
+7  = (1, 2) = 10 10
+8  = (1, 3) = 10 110
+9  = (1, 4) = 10 111
+10 = (2, 0) = 110 00
+</pre>
+
+With the two functions above, we're able to efficiently compress a single
+integer using Golomb-Rice coding. In the next section, we'll put everything
+together and use the primitives described above to construct our compact sets.
+
+=== Golomb-Rice Coded Sets ===
+
+Rather than insert items directly into our set, we instead first run the items
+through a PRF. This creates a set of uniformly distributed values. If we then
+sort each of these values, the ''delta'' between each of the values closely
+resembles a ''Geometric Distribution''. We'll again leverage this fact to use
+Golomb-Rice coding to compresses our set by only encoding the ''delta'' between
+two successive elements in the set.
+
+Golomb-Rice coded sets take two parameters: 
+* <code>N</code> the number of items to be inserted into the set
+* <code>P</code> a value which is computed as <code>1/fp</code> where <code>fp</code> is the desired false positive rate. 
+    
+<code>P</code> can also be interpreted as the parameter to our Geometric
+Distribution.  Intuitively, to achieve a false positive rate of 1/32 (1/2^5),
+in a series of queries of items which ''aren't'' in the set, we expect to
+receive a "NO" (false) 32 times, before getting a "YES" (true, our false
+positive). Once again, <code>P</code> MUST be a power of two.
+
+Given these two parameters, we can now construct our set.
+
+==== Set Construction ====
+
+Set construction takes three parameters: <code>N</code>, <code>P</code> and
+<code>L</code>
+* where <code>L</code> is a list of the ''raw'' items we wish to insert into the set
+* the type of <code>L</code> is assumed to be of <code>[]byte</code>
+
+NOTE: <code>P</code> ''must'' be a power of two as we target the specialized case of
+Golomb coding: Golomb-Rice coding.
+
+Using <code>N</code> and <code>P</code> we compute <code>F = N * P</code>
+<code>F</code> constricts the range of the hashed values accordingly in order to
+achieve our desired false positive rate.
+
+In addition, to help optimize the algorithm, we use a fast range algorithm[7],
+multiplying the hashed value by F and taking only the top 64 bits. This fairly
+distributes the values over F without division and can be done with fewer cycles
```

Comment:
Doesn't mention what strategy this algorithm uses fewer cycles than.

Reply:
Elaborated a bit more.